Documents/Basics/Tutorial 02.xml

```diff
@@ -245,8 +245,8 @@ void main()
                 <example>
                     <title>Vertex Arrays</title>
                     <programlisting language="cpp"><![CDATA[GLbyte *bufferObject = (void*){0.0f, 0.5f, 0.0f, 1.0f, 0.5f, -0.366f, ...};
-float *positionAttribArray[4] = (float *[4])(&(bufferObject + 0));
-float *colorAttribArray[4] = (float *[4])(&(bufferObject + 48));]]></programlisting>
+float *positionAttribArray[3] = (float *[4])(&(bufferObject + 0));

```

Comment:
> Shouldn't positionAttribArray and colorAttribArray have 3 elements? That's what I've changed it to

yes, this is wrong. The code states
> ptr to 4D element

you changed it to

> ptr to 3D element

the number of elements is unspecified there.

Reply:
Hmm I'm still confused because in the next pseudo code example you write
```
void glDrawArrays(GLenum type, GLint start, GLint count)
{
    for(GLint element = start; element < start + count; element++)
    {
        VertexShader(positionAttribArray[element], colorAttribArray[element]);
    }
}
```
and from the tutorial we call
`glDrawArrays(GL_TRIANGLES, 0, 3);`
so doesn't that result in
```
VertexShader(positionAttribArray[0], colorAttribArray[0])
VertexShader(positionAttribArray[1], colorAttribArray[1])
VertexShader(positionAttribArray[2], colorAttribArray[2])
```
I guess I just don't know what this line is doing
`float *positionAttribArray[4] = (float *[4])(&(bufferObject + 0));`
Is *float \*positionAttribArray[4]* not an array with 4 elements? I tried recreating this code in a simple program but I couldn't get it to compile. I've studied pointer arithmetic so is it something like
```
positionAttribArray[0] = (positionAttribArray + 0) = bufferObject[0-15]
positionAttribArray[1] = (positionAttribArray + 16) = bufferObject[16-31]
positionAttribArray[2] = (positionAttribArray + 32) = bufferObject[32-47]
```
If that's right then it's this specific bit I don't get
`(float *[4])(&(bufferObject + 0))`
and in the line before this one we have a void pointer to 48
`glVertexAttribPointer(1, 4, GL_FLOAT, GL_FALSE, 0, (void*)48);`
Does this tutorial assume a float size of 4 bytes? Because future tutorials always use sizeof(float). Sorry for this not being OpenGL related at all. I'm going to blame the book I learned C++ from for not preparing me for this pointer stuff.

If the example in the tutorial is correct I'll file this away for later study

Reply:
> and from the tutorial we call
glDrawArrays(GL_TRIANGLES, 0, 3);
so doesn't that result in

this is correct. That pointer part is nothing special actually (besides being pseudocode). What it wants to declare is:

```
float4 positionAttribArray[]
```
but there is no float4 in C++, so we write
```
float positionAttribArray[][4]
```
but we cant have an empty dimension at the declaration, so we write
```
float *positionAttribArray[4]
```

Reply:
I've removed this commit so it should be good to go! Thanks for your patience with me while I figure all this out